### PR TITLE
MRG+1: Add append channels raw

### PR DESCRIPTION
**Pull request by choldgraf**
While writing the tutorial for how to concatenate channels across `Raw` objects, we realized that there actually wasn't a way to concatenate channels across `Raw` objects. So here's a quick method to do so using the _merge_info function and some quick checks to make sure that data makes sense to merge. Thoughts @wmvanvliet @agramfort @dengemann @Eric89GXL?

### CODE REVIEW

File: mne/io/base.py [hunk 1]
```diff
@@ -1629,6 +1629,33 @@ def append(self, raws, preload=None):
...
+        """
+        from .array import RawArray
+        if not all([raw.preload for raw in raw_list]):
+            raise AssertionError('All data must be preloaded')
+
+        if not all([self.n_times == raw.n_times for raw in raw_list]):
+            raise AssertionError('Data have unequal numbers of times')
+
+        # Pull data from each raw object to be appended, then vstack
+        infos, datas = zip(*[(raw.info, raw._data) for raw in raw_list])
```
Comment (Eric89GXL) [inline mne/io/base.py]: testing will need to be much more extensive, e.g. some can be preloaded, others not, what to do in those situations

File: mne/io/fiff/tests/test_raw.py [hunk 2]
```diff
@@ -115,6 +115,19 @@ def test_copy_append():
...
+def test_append_channels():
+    """Test raw splitting / re-appending channel types
+    """
+    raw = Raw(fif_fname, preload=True)
+    raw_eeg = raw.pick_types(eeg=True, meg=False)
+    raw_meg = raw.pick_types(meg=True, eeg=False)
+    raw_new = raw_meg.append_channels([raw_meg])
+    assert_true([ch in raw_new.ch_names for ch in raw.ch_names])
+    assert_true(raw_new.n_times == raw.n_times)
+    assert_true(len(raw_new.ch_names) == len(raw.ch_names))
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: you should actually check that `raw[:366, :]` times and data are equal to `raw_new[:, :]`

**Comment by Eric89GXL** [comment]
Looks like a reasonable start to me

File: mne/io/base.py (hunk 1 above)
Comment (choldgraf) [inline mne/io/base.py]: As in, some Raw objects cannot be preloaded because they're too big?

Comment (Eric89GXL) [inline mne/io/base.py]: Ohh never mind, you check for the preload state above, I missed that

File: mne/io/fiff/tests/test_raw.py (hunk 2 above)
Comment (choldgraf) [inline mne/io/fiff/tests/test_raw.py]: See latest commit - that work for you?

**Comment by choldgraf** [comment]
Think that I should add a quick check at the beginning `if isinstance(raw_lists, _BaseRaw): raw_lists = list(raw_lists)` or something like this? Or keep it the way it is?

**Comment by Eric89GXL** [comment]
I think you should just raise an exception if it's not a list or tuple

Also, please add `assert_raises` for the things that should raise exceptions

File: mne/io/fiff/tests/test_raw.py [hunk 3]
```diff
@@ -115,6 +115,33 @@ def test_copy_append():
...
+
+    # Now test errors
+    raw_badsf = raw_eeg.copy()
+    raw_badsf.info['sfreq'] = 3.1415927
+    raw_eeg = raw_eeg.crop(5)
+
+    assert_raises(AssertionError, raw_meg.append_channels, [raw_nopre])
+    assert_raises(RuntimeError, raw_meg.append_channels, [raw_badsf])
+    assert_raises(AssertionError, raw_meg.append_channels, [raw_eeg])
+    assert_raises(ValueError, raw_meg.append_channels, [raw_meg])
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: oh you already have them now, cool :)

Comment (choldgraf) [inline mne/io/fiff/tests/test_raw.py]: :)

File: mne/io/fiff/tests/test_raw.py [hunk 4]
```diff
@@ -115,6 +115,33 @@ def test_copy_append():
...
 
 
+@testing.requires_testing_data
+def test_append_channels():
+    """Test raw splitting / re-appending channel types
+    """
+    raw = Raw(fif_fname, preload=True)
+    raw_nopre = Raw(fif_fname, preload=False)
+    raw_eeg = raw.pick_types(meg=False, eeg=True, copy=True)
+    raw_meg = raw.pick_types(meg=True, eeg=False, copy=True)
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: you could also do `raw_stim = raw.pick_types(meg=False, eeg=False, stim=True, copy=True)` and check that none of those channels are in your `raw_new`

File: mne/io/fiff/tests/test_raw.py [hunk 5]
```diff
@@ -115,6 +115,33 @@ def test_copy_append():
...
+    raw_new = raw_meg.append_channels([raw_eeg])
+    nchans = len(raw_new.ch_names)
+
+    assert_true([ch in raw_new.ch_names for ch in raw.ch_names])
+    assert_true(raw_new[:][0].shape == raw[:nchans][0].shape)
+    assert_array_equal(raw_new[0, :][1], raw[0, :][1])
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: What I meant was:
```
assert_array_equal(raw_new[:, :][0], raw[:nchans, :][0])
assert_array_equal(raw_new[:, :][1], raw[:, :][1])
```

File: mne/io/fiff/tests/test_raw.py [hunk 6]
```diff
@@ -115,6 +115,33 @@ def test_copy_append():
...
+    """
+    raw = Raw(fif_fname, preload=True)
+    raw_nopre = Raw(fif_fname, preload=False)
+    raw_eeg = raw.pick_types(meg=False, eeg=True, copy=True)
+    raw_meg = raw.pick_types(meg=True, eeg=False, copy=True)
+    raw_new = raw_meg.append_channels([raw_eeg])
+    nchans = len(raw_new.ch_names)
+
+    assert_true([ch in raw_new.ch_names for ch in raw.ch_names])
+    assert_true(raw_new[:][0].shape == raw[:nchans][0].shape)
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: and then you don't even need this check because it's done essentially by the `assert_array_equal` below

File: mne/io/fiff/tests/test_raw.py [hunk 7]
```diff
@@ -115,6 +115,33 @@ def test_copy_append():
     assert_equal(data.shape[1], 2 * raw._data.shape[1])
 
 
+@testing.requires_testing_data
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: you also don't need the testing data for this, use `test_fif_fname` instead, that's included with the repo

File: mne/io/fiff/tests/test_raw.py [hunk 8]
```diff
@@ -115,6 +115,33 @@ def test_copy_append():
...
+def test_append_channels():
+    """Test raw splitting / re-appending channel types
+    """
+    raw = Raw(fif_fname, preload=True)
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: and for speed, you should do `raw = Raw(fif_fname).crop(0, 1).preload_data()`

Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: which reminds me, what happens if I try to channel-concatenate raws with different numbers of time points? There should be an error. I think right now it is cryptic, you should add a check to make sure it's informative.

what happens if I try to channel-concatenate raws with different sampling rates, but the same number of time points?

what happens if I try to channel-concatenate raws with `.info` with e.g. different subject names?

etc.

**Comment by choldgraf** [comment]
I think I addressed all of your comments - lemme know what you think

**Comment by choldgraf** [comment]
Oops, missed a few comments, inline responses to follow

File: mne/io/fiff/tests/test_raw.py [hunk 9]
```diff
@@ -115,6 +115,34 @@ def test_copy_append():
...
+    raw_eeg_meg = raw.pick_types(meg=True, eeg=True, copy=True)
+    raw_eeg = raw.pick_types(meg=False, eeg=True, copy=True)
+    raw_meg = raw.pick_types(meg=True, eeg=False, copy=True)
+    raw_stim = raw.pick_types(meg=False, eeg=False, stim=True, copy=True)
+    raw_new = raw_meg.append_channels([raw_eeg])
+
+    assert_true([ch in raw_new.ch_names for ch in raw.ch_names])
+    assert_array_equal(raw_new[:, :][0], raw_eeg_meg[:, :][0])
+    assert_array_equal(raw_new[:, :][1], raw[:, :][1])
+    assert_true(all([ch not in raw_new.ch_names for ch in raw_stim.ch_names]))
```
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: FYI you do not need the `[]` here. There is actually a difference -- with the `[]`, the whole list is generated, then `all` evaluated, whereas without it, `all` does the iterating, and therefore will short-circuit on the first non-`True` entry if it can. Thus it is more efficient not to have it. Here it doesn't matter because the computation is so simple with so few entries, but other times it can matter for efficiency.

Comment (agramfort) [inline mne/io/fiff/tests/test_raw.py]: is this a MRG or WIP PR?

File: mne/io/fiff/tests/test_raw.py (hunk 8 above)
Comment (choldgraf) [inline mne/io/fiff/tests/test_raw.py]: > what happens if I try to channel-concatenate raws with different numbers of time points? 
```python
if not all([self.n_times == raw.n_times for raw in raw_list]):
            raise AssertionError('Data have unequal numbers of times')
```
this is already in there and does it, no?

>what happens if I try to channel-concatenate raws with different sampling rates, but the same number of time points?

_merge_info raises a decent error if the sampling frequencies don't match up

> what happens if I try to channel-concatenate raws with .info with e.g. different subject names?

Right now it does not raise any errors for this kind of discrepancy, but this should be the job of `_merge_info` rather than the `append_channels` function, no?

File: mne/io/fiff/tests/test_raw.py (hunk 9 above)
Comment (choldgraf) [inline mne/io/fiff/tests/test_raw.py]: I think MRG, no? I mean, I would like it to be MRG :)

File: mne/io/fiff/tests/test_raw.py (hunk 8 above)
Comment (Eric89GXL) [inline mne/io/fiff/tests/test_raw.py]: Oh good, I wasn't sure if merge info was smart or not

**Comment by choldgraf** [comment]
Looks like all of the builds passed except for one in Travis that returned a missing data error. Is that something that I have the ability to fix? If not, then I can add this to "what's new" and the API docs. Then ready to merge?

**Comment by Eric89GXL** [comment]
You need to actually use the different filename that I mentioned in a previous comment

**Comment by Eric89GXL** [comment]
We have testing files as part of the repo, that is what you should use here. We also have testing files as part of the testing dataset that are overkill for what you ate doing here. You removed the testing decorator but still attempt to use the testing file instead of the repo file, make sense?

**Comment by choldgraf** [comment]
Ah I think that I changed this locally but didn't push it to the server,
will do first thing tomorrow.

**Comment by choldgraf** [comment]
Mmk updated the test file name as well as changelog. I should update the API documentation as well, correct?

**Comment by choldgraf** [comment]
Looks like builds pass and docstring updated - ready to merge?

File: mne/io/base.py [hunk 10]
```diff
@@ -1629,6 +1629,36 @@ def append(self, raws, preload=None):
...
+
+        if not all(self.n_times == raw.n_times for raw in raw_list):
+            raise AssertionError('Data have unequal numbers of times')
+
+        # Pull data from each raw object to be appended, then vstack
+        infos, datas = zip(*[(raw.info, raw._data) for raw in raw_list])
+        data = np.vstack([self._data] + list(datas))
+        new_info = _merge_info([self.info] + list(infos))
+        new_raw = RawArray(data, new_info)
+        return new_raw
```
Comment (dengemann) [inline mne/io/base.py]: should mofify inplace, think about sustainable naming compatible with epochs and evoked for future mixin

**Comment by choldgraf** [comment]
OK so short comment == large change :)

Now it's not a method of Raw only, it's a Mixin. I've renamed the PickDropChannelsMixin to AddPickDropChannelsMixin (I think that's it). The tests that I had before pass, and now there should be functionality for adding channels with an `add_channels` method for `Raw, Epochs, Evoked, and AverageTFR`. I need to write more tests for these other versions, but want to see what people think first.

For tests - one big "test APDCMixin" function, or one separate function for each class type that inherit it?

File: mne/io/base.py [hunk 11]
```diff
@@ -1629,6 +1629,44 @@ def append(self, raws, preload=None):
...
+        # Pull data from each raw object to be appended, then vstack
+        infos, datas = zip(*[(raw.info, raw._data) for raw in raw_list])
+        data = np.vstack([self._data] + list(datas))
+        new_info = _merge_info([self.info] + list(infos))
+        if copy is True:
+            new_raw = RawArray(data, new_info)
+        else:
+            self._data = data
+            self.info = new_info
+            new_raw = self
```
Comment (dengemann) [inline mne/io/base.py]: this can go away now!

File: doc/source/whats_new.rst [hunk 12]
```diff
@@ -31,6 +31,9 @@ Changelog
 
     - Add new object :class:`mne.decoding.TimeDecoding` for decoding sensors' evoked response across time by `Jean-Remi King`_
 
+    - Add support to append new channels from a list of Raw objects by `Chris Holdgraf`_
```
Comment (agramfort) [inline doc/source/whats_new.rst]: as you modified the mixin you added the feature to all objects (Epochs, Evoked etc...). Although only Raw is tested.

Comment (dengemann) [inline doc/source/whats_new.rst]: Good point, let's test it before merging it.

Comment (choldgraf) [inline doc/source/whats_new.rst]: Planning to do the rest of tests tomorrow... Should it be one function that
tests all the class types, or a separate "test_channel_update" test for
each type?

Comment (agramfort) [inline doc/source/whats_new.rst]: I would use a different test function for each file type.

Comment (dengemann) [inline doc/source/whats_new.rst]: +1

Comment (choldgraf) [inline doc/source/whats_new.rst]: Hey all - I'm finally done being super sick and would like to finish up
this PR. I'm writing tests for Raw/Epochs/AverageTFR/Evoked, but had a
question to run past you guys:

For Evoked, I ran into the same problem with a missing field in "info" that
I ran into when creating RawArray objects using "create_info". In that
case, the field "buffer_size_sec" was missing when you created a new info
object from scratch. When you tried to merge Info objects with _merge_info,
it expected that field and threw a KeyError. So, I just included that in
the set of fields and it worked fine.

Now I am reading in the test evoked data like this:

`evoked = read_evokeds(fname, condition=0)`

And I'm getting the same error. So I wanted to ask you guys: is
buffer_size_sec something that should *not* be in the info objects? Instead
of adding the field to the `create_info` function, should I have instead
removed that field from the `_merge_info` function?

Comment (wmvanvliet) [inline doc/source/whats_new.rst]: +1 for adding it to meas_info.py/RAW_INFO_FIELDS. This way, it gets initialised to None with `create_info`.

Comment (agramfort) [inline doc/source/whats_new.rst]: buffer_size_sec  only makes sense for raw indeed.

Comment (choldgraf) [inline doc/source/whats_new.rst]: So adding buffer_size_sec to Raw info fields only won't fix this error, since the evoked object being read in doesn't have a "buffer_size_sec" field and _merge_info checks for it. Is it kosher if I just manually add that field in before running the test? E.g.:

```python
evoked = read_evokeds(fname, condition=0)
evoked.info['buffer_size_sec'] = None
```
(I can confirm that this makes the tests for evoked pass)

Otherwise it seems like we'd need to modify _merge_info to not always check for all of those fields...

Comment (choldgraf) [inline doc/source/whats_new.rst]: And I think that from now on 'buffer_size_sec' should always be in any new info object, since it is now in RAW_INFO_FIELDS, right?

**Comment by choldgraf** [comment]
hrm....this isn't mergeable now, I just tried rebasing but it's still unhappy

**Comment by choldgraf** [comment]
Perhaps this is because of the renaming to "UpdateChannelsMixin"?

**Comment by choldgraf** [comment]
After pulling from upstream/master and testing a merge locally, it worked fined, so I am not sure what is wrong...any ideas?

**Comment by agramfort** [comment]
squash all your commits into one and try to rebase again.

**Comment by agramfort** [comment]
wait it seems you fixed it...

**Comment by choldgraf** [comment]
it still says AppVeyor wasn't able to build a non-mergeable PR...is that not what you're getting?

**Comment by choldgraf** [comment]
OK I think I just smooshed them all into a single commit and force pushed it. Looks like the builds are taking place now, so perhaps it worked. Fingers crossed on the build.

**Comment by Eric89GXL** [comment]
I'd rather modify `_merge_info` to be smarter about what fields it checks than make `epochs.info` dumber about what fields it has :)

**Comment by choldgraf** [comment]
If we're going to go that route, then wouldn't we also want to modify
`create_info` to create different info objects for different purposes?
E.g., if you want to create an info object for an EpochsArray, then you'd
specify that and it would not include the "buffer_size_sec" key.

**Comment by Eric89GXL** [comment]
Maybe, but we might not need it, either. The idea being that in most cases it shouldn't matter if I have *all* the fields an object supports if I'm using an `*Array` type, I just need the ones that will allow me to use my data appropriately. And omitting `buffer_size_sec` I should still allow RawArray to work, so we can just omit that field for now from all types, which I think we already do.

**Comment by choldgraf** [comment]
so, completely remove "buffer_size_sec" from RAW_INFO_FIELDS and from
create_info, and then make `_merge_infos` check to see if it's there before
doing the merge rather than assuming it is there and breaking when it isn't?

**Comment by Eric89GXL** [comment]
Right now in `master` `buffer_size_sec` isn't in `RAW_INFO_FIELDS`, and it's not put in by `create_info`, which seems reasonable to me. We should probably rename `RAW_INFO_FIELDS` to e.g. `_INFO_FIELDS` or something. And then think of `buffer_size_sec` as a field that should probably be there for raw, but might not be. And `_merge_info` can treat it in a special way, since it probably should -- make sure all infos have it and they all match, or that none of them have it.

**Comment by choldgraf** [comment]
ok cool - I mentioned removing them because they're current slated to be
merged in this PR :)

OK so I will undo those changes and tell _merge_info to check for
buffer_size_sec

**Comment by wmvanvliet** [comment]
I've been documenting the info object and found that all info fields at the moment are required, except `proc_history` (which is currently the only exception, this should be addressed). Many of them can be set to `None`, but the code always assumes the field to be there. I am all for simplifying the info object, but this is a huge undertaking. At the moment, simply adding `buffer_size_sec` to RAW_INFO_FIELDS is the best solution.

**Comment by Eric89GXL** [comment]
@wmvanvliet but that field is not required all the time. Do you mean we should add it and set it to `None`?

**Comment by wmvanvliet** [comment]
yes, that's how it works with the other fields as well.

**Comment by Eric89GXL** [comment]
Ahh, I guess that's fine, since nothing will be written out if files are saved with `info['buffer_size_sec'] = None`

**Comment by wmvanvliet** [comment]
yes.

**Comment by choldgraf** [comment]
OK cool - so in that case what's left to be done on this PR? Looks like travis wasn't happy because I forgot to make flake, but should be good to go now.

**Comment by Eric89GXL** [comment]
Looks like you need a rebase

**Comment by Eric89GXL** [comment]
Otherwise LGTM, +1 for merge once rebase is done

**Comment by choldgraf** [comment]
cool - I just fixed it, the rebase was just for `whats_new`

**Comment by Eric89GXL** [comment]
Great, I'll let someone else take a final look and merge if they're happy

File: mne/tests/test_evoked.py [hunk 13]
```diff
@@ -452,3 +452,33 @@ def test_array_epochs():
     types = ['eeg'] * 19
     info = create_info(ch_names, sfreq, types)
     assert_raises(ValueError, EvokedArray, data1, info, tmin=-0.01)
+
+
+def test_add_channels():
+    """Test evoked splitting / re-appending channel types
+    """
+    evoked = read_evokeds(fname, condition=0)
+    evoked.info['buffer_size_sec'] = None
```
Comment (agramfort) [inline mne/tests/test_evoked.py]: will this be needed by user saving there evoked files on disk?

Comment (choldgraf) [inline mne/tests/test_evoked.py]: That's a good question. Right now it populates that field when creating any new info object. I had never run into this problem before trying to get this test to pass. Has anyone else had it before?

Comment (choldgraf) [inline mne/tests/test_evoked.py]: Hmmm, I just tried creating an EvokedArray, confirmed that "buffer_size_sec" was in there, then wrote to disk and read it back in, and the resulting object no longer had that field. Maybe this would end up being a problem? Either way it sounds like there should be more consistency on the part of this attribute...

Comment (agramfort) [inline mne/tests/test_evoked.py]: that's a bug in the evoked IO code that needs to be fixed in another PR...

grrrr

Comment (choldgraf) [inline mne/tests/test_evoked.py]: ah ok - so then are we good on this PR?

**Comment by choldgraf** [comment]
can I check this off of my to-do list? :)

**Comment by choldgraf** [comment]
spoke too soon, looks like I needed another rebase...hopefully it's ready now?

**Comment by dengemann** [comment]
Going over it.

**Comment by dengemann** [comment]
Thanks @choldgraf !

**Comment by choldgraf** [comment]
Wahoo!
